internal/framework/status/updater.go

```diff
@@ -38,56 +47,93 @@ type UpdaterConfig struct {
 // status API calls sequentially will take time.
 // (b) k8s API can become slow or even timeout. This will increase every update status API call.
-// Making updaterImpl asynchronous will prevent it from adding variable delays to the event loop.
+// Making UpdaterImpl asynchronous will prevent it from adding variable delays to the event loop.
 //
-// (4) It doesn't retry on failures. This means there is a chance that some resources will not have up-to-do statuses.
+// (3) It doesn't retry on failures. This means there is a chance that some resources will not have up-to-do statuses.
 // Statuses are important part of the Gateway API, so we need to ensure that the Gateway always keep the resources
 // statuses up-to-date.
 //
-// (5) It doesn't clear the statuses of a resources that are no longer handled by the Gateway. For example, if
+// (4) It doesn't clear the statuses of a resources that are no longer handled by the Gateway. For example, if
 // an HTTPRoute resource no longer has the parentRef to the Gateway resources, the Gateway must update the status
 // of the resource to remove the status about the removed parentRef.
 //
-// (6) If another controllers changes the status of the Gateway/HTTPRoute resource so that the information set by our
+// (5) If another controllers changes the status of the Gateway/HTTPRoute resource so that the information set by our
 // Gateway is removed, our Gateway will not restore the status until the EventLoop invokes the StatusUpdater as a
 // result of processing some other new change to a resource(s).
 // FIXME(pleshakov): Make updater production ready
 // https://github.com/nginxinc/nginx-kubernetes-gateway/issues/691
 
-// To support new resources, updaterImpl needs to be modified. Consider making updaterImpl extendable, so that it
+// UpdaterImpl needs to be modified to support new resources. Consider making UpdaterImpl extendable, so that it
 // goes along the Open-closed principle.
-type updaterImpl struct {
-	cfg UpdaterConfig
+type UpdaterImpl struct {
+	leaderElector LeaderElector
+	lastStatuses  *Statuses
+	cfg           UpdaterConfig
+
+	statusLock sync.Mutex
 }
 
 // NewUpdater creates a new Updater.
-func NewUpdater(cfg UpdaterConfig) Updater {
-	return &updaterImpl{
+func NewUpdater(cfg UpdaterConfig) *UpdaterImpl {
+	return &UpdaterImpl{
 		cfg: cfg,
 	}
 }
 
-func (upd *updaterImpl) Update(ctx context.Context, statuses Statuses) {
+// SetLeaderElector sets the LeaderElector of the updater.
+func (upd *UpdaterImpl) SetLeaderElector(elector LeaderElector) {
+	upd.leaderElector = elector
+}
+
+// WriteLastStatuses writes the last saved statuses for the Gateway API resources.
+// Used in leader election when the Pod starts leading. It's possible that during a leader change,
+// some statuses are missed. This will ensure that the latest statuses are written when a new leader takes over.
```

Comment:
just noticed an edge case, because the update to the config resource is processed separately from the graph updates -- https://github.com/nginxinc/nginx-kubernetes-gateway/blob/7871727d60a1aff70738827250f9b06a33d998a3/internal/mode/static/handler.go#L73 -- the last status might not have the statuses of the resources. This means  "This will ensure that the latest statuses are written when a new leader takes over." will not always work.

Reply:
Good catch. That's a problem.

Any reason we can't write the config resource status with the rest of the statuses at the end of HandleEventBatch? cc @sjberman

Reply:
I guess it could work that way...I think I implemented it to keep the separation since a control config update happens separately from Gateway updates. Maybe there's a technical reason I'm not remembering, but worth a shot to combine them.

Reply:
This approach doesn't fix the problem. If the event batch contains a config resource update but does not change the nginx config, then we end up sending a status update with just the NginxGatewayStatus set, which effectively clears the last saved statuses for the Gateway API resources. 

Without taking on a big refactor of the Status Updater, I think we have the following options:
1. Build the statuses of the Gateway API resources even when the event batch does not contain any changes to the NGINX config. This means that we may end up re-writing Gateway API statuses.
2. Infer what type of status update we are performing in the Updater by checking if the NginxGatewayStatus is nil. If it's nil, then we are updating Gateway API resources and we only save those fields. If it is not nil, then we are only updating the NginxGateway status and we only save that off that field. This assumes that we will never update both the Gateway API statuses and the NginxGateway status in the same call to update.
3. Refactor Update to accept an interface that's implemented by both NginxGatewayStatus and Statuses (will be changed to only contain Gateway API statuses). Updater determines what type of update it is performing by switching on the interface type. 
4. Add a new method to the Updater that updates the NginxGateway status. The other Update method will only update the status for Gateway API resources.

I'm leaning towards option 3, but I think 4 wouldn't be all that bad either. 

@pleshakov @sjberman @ciarams87 thoughts or other ideas?

Reply:
4 would probably involve some duplication, no?

3 sounds good in theory

Reply:
> 4 would probably involve some duplication, no?
Not really. We'd just pull out the call to update the status of the NginxGateway resource into its own method.

Reply:
> 3. Refactor Update to accept an interface that's implemented by both NginxGatewayStatus and Statuses (will be changed to only contain Gateway API statuses). Updater determines what type of update it is performing by switching on the interface type.

Removing  status updater of any knowledge of _specific_ resource types makes a lot of sense to me - this way it can be used update status of any resource.

However, that'll probably mean we will need to accumulate the statuses in the handler, and the updater needs to not update statuses if they don't change (for performance optimization). This also means the handler needs to be aware of leader election, not the status updater (crazy idea: Perhaps we can even introduce a new type of an event specific to leader election)

Reply:
> Removing status updater of any knowledge of specific resource types makes a lot of sense to me - this way it can be used update status of any resource.

Makes a lot of sense to me too, but that would be a larger refactor.

Reply:
Implemented option 3: https://github.com/nginxinc/nginx-kubernetes-gateway/pull/1048/commits/abd6099a3c5d5e289fd82bb123b654c58c150bfe

Reply:
> Makes a lot of sense to me too, but that would be a larger refactor.
would it make sense to add it as a tech debt issue?

Reply:
Sure, I can add.

Reply:
https://github.com/nginxinc/nginx-kubernetes-gateway/issues/1071